### PR TITLE
syncer: split functions of different shard mode

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
handleQueryEvent is too big 😂 

### What is changed and how it works?
depends on different shard mode, program will go to `handleQueryEventNoSharding`, `handleQueryEventOptimistic`, `handleQueryEventPessimistic`

### Check List 

Tests 

 - pass original tests

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -2503,8 +2533,8 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 			return err
 		}
 		// maybe multi-groups' sharding DDL synced in this for-loop (one query-event, multi tables)
-		if cap(*ec.shardingReSyncCh) < len(sqls) {
-			*ec.shardingReSyncCh = make(chan *ShardingReSync, len(sqls))
+		if cap(*ec.shardingReSyncCh) < len(needHandleDDLs) {
```
Comment (lance6716) [inline syncer/syncer.go:2536]: in fact I think this part is not used now 🤔

File: syncer/syncer.go [hunk 2]
```diff
@@ -2364,74 +2364,104 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
-		})
+	switch s.cfg.ShardMode {
+	case "":
+		return s.handleQueryEventNoSharding(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL, sourceTbls)
+	case config.ShardOptimistic:
+		return s.handleQueryEventOptimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL)
+	case config.ShardPessimistic:
+		return s.handleQueryEventPessimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL, ddlInfo)
+	}
+	return fmt.Errorf("unsupported shard-mode %s, should not happened", s.cfg.ShardMode)
```
Comment (GMHDBJD) [inline syncer/syncer.go:2375]: errors.Errorf ?

**Review by GMHDBJD — APPROVED** [review]
rest LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

**Review by lichunzhu — APPROVED** [review]

File: syncer/syncer.go [hunk 3]
```diff
@@ -2364,74 +2364,104 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
-		})
+	switch s.cfg.ShardMode {
+	case "":
+		return s.handleQueryEventNoSharding(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL, sourceTbls)
+	case config.ShardOptimistic:
+		return s.handleQueryEventOptimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL)
+	case config.ShardPessimistic:
+		return s.handleQueryEventPessimistic(ev, ec, needHandleDDLs, needTrackDDLs, onlineDDLTableNames, originSQL, ddlInfo)
+	}
+	return errors.Errorf("unsupported shard-mode %s, should not happened", s.cfg.ShardMode)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply syncer/syncer.go:2375]: terror need we assign an error code, but this error is just a marker, it will not happen

**Pull request merged**